pkg/cmd/server/kubernetes/node/options/options.go

```diff
@@ -3,178 +3,152 @@ package node
-		Webhook: kubeletconfig.KubeletWebhookAuthorization{
-			CacheAuthorizedTTL:   metav1.Duration{Duration: authzTTL},
-			CacheUnauthorizedTTL: metav1.Duration{Duration: authzTTL},
-		},
+	if network.IsOpenShiftNetworkPlugin(options.NetworkConfig.NetworkPluginName) {
+		// SDN plugin pod setup/teardown is implemented as a CNI plugin
+		setIfUnset(args, "network-plugin", kubeletcni.CNIPluginName)
+		setIfUnset(args, "cni-conf-dir", kubeletcni.DefaultNetDir)
+		setIfUnset(args, "cni-bin-dir", kubeletcni.DefaultCNIDir)
+		setIfUnset(args, "hairpin-mode", kubeletconfig.HairpinNone)
+	} else {
+		setIfUnset(args, "network-plugin", options.NetworkConfig.NetworkPluginName)
 	}
 
-	// resolve extended arguments
-	// TODO: this should be done in config validation (along with the above) so we can provide
-	// proper errors
-	if err := cmdflags.Resolve(options.KubeletArguments, server.AddFlags); len(err) > 0 {
-		return nil, kerrors.NewAggregate(err)
-	}
+	// prevents kube from generating certs
+	setIfUnset(args, "tls-cert-file", options.ServingInfo.ServerCert.CertFile)
+	setIfUnset(args, "tls-private-key-file", options.ServingInfo.ServerCert.KeyFile)
+	// roundtrip to get a default value
+	setIfUnset(args, "tls-cipher-suites", crypto.CipherSuitesToNamesOrDie(crypto.CipherSuitesOrDie(options.ServingInfo.CipherSuites))...)
+	setIfUnset(args, "tls-min-version", crypto.TLSVersionToNameOrDie(crypto.TLSVersionOrDie(options.ServingInfo.MinTLSVersion)))
 
-	// terminate early if feature gate is incorrect on the node
-	if len(server.FeatureGates) > 0 {
-		if err := utilfeature.DefaultFeatureGate.SetFromMap(server.FeatureGates); err != nil {
+	// Server cert rotation is ineffective if a cert is hardcoded.
+	if len(args["feature-gates"]) > 0 {
+		// TODO this affects global state, but it matches what happens later.  Need a less side-effecty way to do it
+		if err := utilfeature.DefaultFeatureGate.Set(args["feature-gates"][0]); err != nil {
 			return nil, err
 		}
-	}
-	if utilfeature.DefaultFeatureGate.Enabled(features.RotateKubeletServerCertificate) {
-		// Server cert rotation is ineffective if a cert is hardcoded.
-		if len(server.CertDirectory) > 0 {
-			server.TLSCertFile = ""
-			server.TLSPrivateKeyFile = ""
+		if utilfeature.DefaultFeatureGate.Enabled(features.RotateKubeletServerCertificate) {
+			// Server cert rotation is ineffective if a cert is hardcoded.
+			setIfUnset(args, "tls-cert-file", "")
+			setIfUnset(args, "tls-private-key-file", "")
 		}
 	}
 
-	if network.IsOpenShiftNetworkPlugin(options.NetworkConfig.NetworkPluginName) {
-		// SDN plugin pod setup/teardown is implemented as a CNI plugin
-		server.NetworkPluginName = kubeletcni.CNIPluginName
-		server.CNIConfDir = kubeletcni.DefaultNetDir
-		server.CNIBinDir = kubeletcni.DefaultCNIDir
-		server.HairpinMode = kubeletconfig.HairpinNone
-	}
-
-	server.RootDirectory, err = filepath.Abs(server.RootDirectory)
+	// we sometimes have different clusterdns options.  I really don't understand why
```

Comment:
This is not correct, this will cause us to not be using the correct DNS server (it regresses us pulling from the nodes).

This value needs to be set based on what is in node-config, and the optional resolution needs to happen.

We can't ship this.

Reply:
Edit, I tracked down what was happening - it was specifically a bootstrap mode thing, where the client can't exist until the bootstrap config does.  This has to be an optional block (we can't error out if the master kubeconfig doesn't exist) because sometimes we only have bootstrap-kubeconfig.  So the copied logic is ok, but not the failing if masterkubeconfig doesn't exist yet (which it won't when bootstrap-kubeconfig is set).

Reply:
Actually... looking at getClusterDNS, I don't see how it is correct.  We should be defaulting the cluster DNS string, and only if we have no value should we be falling back to cluster DNS.  Current code looks like it always stomps cluster DNS which is wrong when the master sets it.

Reply:
Logic should be

1. if cluster DNS is empty or 0.0.0.0, do a lookup
2. if the lookup succeeded, stomp the current value
3. otherwise do nothing